21 update resource class

vonfredd opened the pull request:
### Related Issue
Closes #21

### Proposed Changes
We have no useful endpoints in our resource class.
We should implement REST CRUD functionality to the class in order to make better use of our application.
A namechange on the class "HelloResource" would be good.

- **Change 1**: GET with id path to find entity of an ID.
- **Change 2**: GET to find all entities.
- **Change 3**: POST to add object.
- **Change 4**: PUT with an id path to update object with parameter data.
- **Change 5**: DELETE with an id path.
- **Change 6**: MovieRepository.deleteMovie now takes id as input (postman wanted id to be able to use it).
- **Change 7**: Added tests to check the responsecodes.
- **Change 8**: Refactored MovieRepositoryTest to use new parameters.
- **Change 9**: Added two dependencies. Resteasy-core to use Dispatcher and MockHttp. Also added resteasy-jackson2-provider to handle JSON input and read in tests.

### Description
Updated old resource class to be of more use for our application. Renamed class and added REST endpoints. Added a MovieResourceTest class to check the responsecodes that the methods return. Made changes to the MovieRepository class that led to changes in MovieRepositoryTest class.

Ebyrdeu commented on src/main/java/dev/ebyrdeu/movie_jakarta/resource/MovieResource.java:
```diff
@@ -0,0 +1,80 @@
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Movies all() {
+        return new Movies(movieRepository.findAll().stream().map(MovieDto::map).collect(Collectors.toList()));
```
```suggestion
    return new Movies(movieRepository.findAll().stream().map(MovieDto::map).toList());
```

Ebyrdeu commented on src/main/java/dev/ebyrdeu/movie_jakarta/resource/MovieResource.java:
```diff
@@ -0,0 +1,80 @@
+    @Path("/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response one(@PathParam("id") @NotNull UUID id) {
+        Optional<Movie> optionalMovie = movieRepository.findById(id);
+        if (optionalMovie.isPresent()) {
+            MovieDto movieDto = MovieDto.map(optionalMovie.get());
+            return Response.ok(movieDto).build();
+        } else {
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
```
```suggestion
     Movie movie = movieRepository.findById(id).orElseThrow(() -> new NotFoundException("Cannot find movie with id: " + id));
        MovieDto movieDto = MovieDto.map(movie);
        return Response.ok(movieDto).build();
```

Ebyrdeu commented on src/main/java/dev/ebyrdeu/movie_jakarta/resource/MovieResource.java:
```diff
@@ -0,0 +1,80 @@
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response add(@NotNull MovieDto movieDto) {
+        Movie movie = movieRepository.saveMovie(MovieDto.map(movieDto));
+        return Response.created(URI.create("http://localhost:8080/app/api/movies/" + movie.getId())).build();
```
```suggestion
              movieRepository.saveMovie(MovieDto.map(movieDto));
        return Response.status(201).build();
```

Ebyrdeu commented on src/main/java/dev/ebyrdeu/movie_jakarta/resource/MovieResource.java:
```diff
@@ -0,0 +1,80 @@
+    @Path("/{id}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response update(@PathParam("id") @NotNull UUID id, @NotNull MovieDto movieDetails) {
+        try{
+            movieRepository.updateMovie(id, MovieDto.map(movieDetails));
+        }catch (NotFoundException e){
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+        return Response.ok().build();
```
```suggestion
   movieRepository.updateMovie(id, MovieDto.map(movieDetails));
        return Response.ok().build();
```

Ebyrdeu replied:
That's actually a good point and it would indeed work. The main issue for me was because we already use a `NotFoundException` when calling both the update and delete methods in the Repository layer. This seemed redundant, and I initially thought something wasn't functioning correctly, prompting you to implement an additional layer of protection.

vonfredd replied:
I was just focused on catching the thrown exception from the called methods. But when the repository throws an exception it still returns the correct response. Should we remove the try-catch? The only thing is that it catches the exception and respond with the same response code that would be returned without the try-catch.

Ebyrdeu replied:
Huh, does it? That feels wrong. It should not... That what I have in my app.
![image](https://github.com/Ebyrdeu/movie_jakarta/assets/42122071/c867ecdd-950f-42d5-b52e-9dee35097e17)

Ebyrdeu commented on src/main/java/dev/ebyrdeu/movie_jakarta/resource/MovieResource.java:
```diff
@@ -0,0 +1,80 @@
+    @DELETE
+    @Path("/{id}")
+    public Response delete(@PathParam("id") @NotNull UUID id) {
+        try {
+            movieRepository.deleteMovie(id);
+        } catch (NotFoundException notFound) {
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+        return Response.noContent().build();
+    }
```
```suggestion
         movieRepository.deleteMovie(id);
        return Response.noContent().build();
```

Ebyrdeu reviewed (changes requested):
Overall, you've done an excellent job. However, I have a few concerns that I'm curious about:

1. You've added two new dependencies. Could you explain what problems they solve?
2. Why are the update and delete operations enclosed in a try-catch block? Was there an issue that made it impossible to implement these operations without it?
3. It appears that you're not really utilizing DTO validation, as you don't seem to use them extensively, and they ultimately don't check anything. Given that they're defined in the DTO, this makes them feel somewhat redundant.

vonfredd commented:
> Overall, you've done an excellent job. However, I have a few concerns that I'm curious about:
> 
> 1. You've added two new dependencies. Could you explain what problems they solve?

Change 9: Added two dependencies. Resteasy-core to use Dispatcher and MockHttp. Also added resteasy-jackson2-provider to handle JSON input and read in tests.

> 2. Why are the update and delete operations enclosed in a try-catch block? Was there an issue that made it impossible to implement these operations without it?

I did not consider it being a checked exception somehow. My bad.

> 3. It appears that you're not really utilizing DTO validation, as you don't seem to use them extensively, and they ultimately don't check anything. Given that they're defined in the DTO, this makes them feel somewhat redundant.

I forgot the Validate annotation..
